### PR TITLE
BUG: Update _svds.py: orthogonalize eigenvectors from arpack to eliminate edge case failures with clustered eigenvalues and update docstring examples

### LINKED ISSUE #11406: scipy.sparse.linalg.svds (v1.4.1) on singular matrix does not give vh vectors in null space
I am trying to use scipy's sparse [svds](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html) to compute the nullspace of a matrix by grabbing the smallest singular values with the `which='SM'` option. However, as shown in the code snippet below, the resulting rows of the `vh`  matrix are not in the null space of the input matrix. This is not a problem in the non-sparse [svd](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.svd.html#scipy.linalg.svd). Any ideas on why this happens?

Reproducing Code ex:
```
import numpy as np
import scipy.sparse as sparse
from scipy.linalg import null_space, svd
from scipy.sparse.linalg import svds

# create a 10x10 singular matrix with a 4-dim null space
dim = 4
size = 10
x = np.random.random((size, size-dim))
y = x[:, :dim] * np.random.random(dim)
mat = np.hstack((x, y))

nz = null_space(mat)
print('Null space vectors:', nz.shape)
assert nz.shape[1] == dim

# use non-sparse svd
u, s, vh = svd(mat)
print(f'Last {dim} singular values are 0:', np.allclose(s[-dim:], 0))
print('Smallest right singular vectors in null space:', np.allclose(mat @ vh[-dim:, :].T, 0))

# Smallest singular values should be 0
sp_mat = sparse.coo_matrix(mat)
su, ss, svh = svds(sp_mat, k=dim, which='SM')
print(f'Last {dim} singular values are 0:', np.allclose(ss, 0))
print('Smallest singular vectors via svds in null space:', np.allclose(sp_mat @ svh.T, 0))
```

Output:
```
Null space vectors: (10, 4)
Last 4 singular values are 0: True
Smallest right singular vectors in null space: True
Last 4 singular values are 0: False
Smallest singular vectors via svds in null space: False
```

Package version info:
scipy: 1.4.1
numpy: 1.16.2

### PR DESCRIPTION
**Pull request by lobpcg**
#### Reference issue
Closes gh-11406, gh-16741. 

#### What does this implement/fix?
As proposed in https://github.com/scipy/scipy/pull/16693#issuecomment-1193369299, copying line 297
eigvec, _ = np.linalg.qr(eigvec)
to before line 337, so that the postprocessing gets orthonormal vectors for arpack as well
This fixes all known issues with non-orthogonal output from ARPACK, including now giving correct both left and right vectors in gh-11406, so this check is added to the unit test. 

As a free bonus, also updated and significantly extended the docsrting example, along the way fixing the bug gh-16741 as a part of it, which has affected every PR test in scipy.scipy (Main refguide_asv_check), e.g., see https://github.com/scipy/scipy/pull/14563#issuecomment-1200290619

#### Additional information
https://github.com/scipy/scipy/pull/16690#issuecomment-1193365725 observes three failures in the Azure jobs for https://github.com/scipy/scipy/pull/16693 prerelease_deps_coverage_64bit_blas, refguide_asv_check and source_distribution. The failure occurs when the test checks that vh @ vh.conj().T is approximately the identity matrix. It turns out that the ARPACK-based SVD solver does not guarantee that this condition is true when the inputs are complex and there is a repeated singular value.

### CODE REVIEW

**Comment by lobpcg** [comment]
The failures appear unrelated to this PR, so ready for review. @WarrenWeckesser ?

**Comment by lobpcg** [comment]
Ready for review. @WarrenWeckesser ?
May be I should add docsting examples to fully address https://github.com/scipy/scipy/issues/16741?

**Comment by lobpcg** [comment]
Reade for review and merge, IMHO.

**Comment by lobpcg** [comment]
While waiting for review, I have significantly extended the docsrting example. Hopefully, this PR gets attention, reviewed and merged... 
There is still another significant issue to fix in svds: https://github.com/scipy/scipy/issues/16330 and one more https://github.com/scipy/scipy/issues/9284 on my potential 2do list but having to wait months for reviews of simple PRs makes it difficult for me to make progress...

**Comment by lobpcg** [comment]
Merging main in https://github.com/scipy/scipy/pull/16712/commits/6c0ce33c407745fb24c3626234105cec3f5281cf for unknown reasons broke the unit test  just for Linux Meson tests / Meson build (3.9, false) and Linux Tests / Python 3.8-dbg with similar assertion errors, e.g., 

__________________ Test_SVDS_LOBPCG.test_small_sigma2[float] ___________________
[gw1] linux -- Python 3.9.13 /opt/hostedtoolcache/Python/3.9.13/x64/bin/python
scipy/sparse/linalg/_eigen/tests/test_svds.py:832: in test_small_sigma2
    assert_allclose(sp_mat.transpose() @ su, 0, atol=1e-5, rtol=1e0)
E   AssertionError: 
E   Not equal to tolerance rtol=1, atol=1e-05
E   
E   Mismatched elements: 40 / 40 (100%)
E   Max absolute difference: 1.86599573
E   Max relative difference: inf
E    x: array([[ 0.203688, -1.720899,  0.80937 ,  0.181332],
E          [ 0.65108 , -1.229791,  1.091333, -0.491714],
E          [ 0.662922, -1.068644,  0.857987,  0.083971],...
E    y: array(0)
        dim        = 4
        dtype      = <class 'float'>
        m          = 10
        mat        = array([[0.75358919, 0.99565464, 0.41765909, 0.84314278, 0.20050386,
        0.2610654 , 0.20873058, 0.16266047, 0.0432..., 0.39618705, 0.43422147, 0.94440681, 0.11885702,
        0.94011348, 0.26331356, 0.06472523, 0.04491724, 0.40666401]])
        n          = 10
        nz         = array([[ 1.57559235e-01, -2.12402361e-01,  1.86718224e-02,
         3.10523488e-02],
       [ 5.70754309e-02,  3.62896...15e-01,
        -9.49710421e-01],
       [ 6.57066007e-01,  5.12042595e-01,  3.81989377e-01,
        -6.11934156e-02]])
        rng        = Generator(PCG64) at 0x7FC7F0F13AC0
        s          = array([4.54984232e+00, 1.21645441e+00, 1.05037971e+00, 9.20361726e-01,
       5.73972786e-01, 4.00007789e-01, 9.63337173e-17, 5.98327270e-17,
       1.76884573e-17, 1.62735323e-18])
        self       = <scipy.sparse.linalg._eigen.tests.test_svds.Test_SVDS_LOBPCG object at 0x7fc81023d220>
        size       = 10
        sp_mat     = <10x10 sparse matrix of type '<class 'numpy.float64'>'
	with 100 stored elements in Compressed Sparse Column format>
        ss         = array([1.17874869e-16, 1.759[863](https://github.com/scipy/scipy/runs/8194816519?check_suite_focus=true#step:13:864)66e-16, 4.61036397e-16, 1.99557031e-15])
        su         = array([[ 0.54511892, -0.39371098, -0.0042301 , -0.48374194],
       [-0.07369829,  0.01159275,  0.07205716, -0.3645317...    [ 0.23624745, -0.326349  , -0.05023194,  0.01401531],
       [-0.42795445, -0.73092631,  0.04194756,  0.17688141]])
        svh        = array([[-1.22886897e-01, -1.21334938e-01, -4.68895602e-02,
         4.69108778e-02, -1.59353418e-17, -4.04233737e-17,
...-3.64019701e-16, -7.33805901e-16,
         7.50998294e-01, -5.83555870e-01,  1.51733636e-01,
        -1.29358556e-01]])
        u          = array([[-0.33397412,  0.30407154, -0.47230311, -0.21496766, -0.14088015,
         0.34489264, -0.57462286, -0.07330383...5883,  0.3167347 , -0.32862208,  0.04004026,
        -0.17628979,  0.23883564,  0.45700088,  0.0643964 , -0.03100367]])
        vh         = array([[-4.40366605e-01, -3.85441693e-01, -3.82074718e-01,
        -4.51394369e-01, -4.15320299e-01, -2.77665105e-01,
... 4.74886893e-18, -3.43984034e-17,
        -1.12109659e-01, -2.61908770e-01, -9.49710421e-01,
        -6.11934156e-02]])
        x          = array([[0.75358919, 0.99565464, 0.41765909, 0.84314278, 0.20050386,
        0.2610654 ],
       [0.14709222, 0.2662157...6076074,
        0.78560949],
       [0.9506525 , 0.39618705, 0.43422147, 0.94440681, 0.11885702,
        0.94011348]])
        y          = array([[0.20873058, 0.16266047, 0.04320398, 0.36305945],
       [0.04074189, 0.04349176, 0.03306512, 0.12853487],
    ...36],
       [0.17[865](https://github.com/scipy/scipy/runs/8194816519?check_suite_focus=true#step:13:866)13 , 0.07[866](https://github.com/scipy/scipy/runs/8194816519?check_suite_focus=true#step:13:867)283, 0.09082089, 0.381[871](https://github.com/scipy/scipy/runs/8194816519?check_suite_focus=true#step:13:872)72],
       [0.26331356, 0.06472523, 0.044[917](https://github.com/scipy/scipy/runs/8194816519?check_suite_focus=true#step:13:918)24, 0.40666401]])

**Comment by lobpcg** [comment]
Could someone please help resolving https://github.com/scipy/scipy/pull/16712#issuecomment-1237401746 ? 
Merging main in https://github.com/scipy/scipy/commit/6c0ce33c407745fb24c3626234105cec3f5281cf gave completely wrong results in the unit test test_small_sigma2 just for Linux Meson tests / Meson build (3.9, false) and Linux Tests / Python 3.8-dbg I have no idea why and how to fix it.

**Comment by rgommers** [comment]
I've seen that `test_small_sigma2` a few more times recently I believe, I think it's a flaky test. A little unclear where it is coming from indeed. Looking at the failures in https://github.com/scipy/scipy/actions/workflows/linux_meson.yml, the first two from other PRs show one `test_small_sigma2` failure as well.

I'll have a closer look at the history here.

**Comment by lobpcg** [comment]
I have added a condition, line 831, to test_svds.py for condition at line 832 to pass in all tests:
        if n < m:  # else the assert fails with some libraries unclear why
            assert_allclose(sp_mat.transpose() @ su, 0, atol=1e-5, rtol=1e0)

I suggest merging it that way and make a new issue on why it fails for future generations to handle.

**Comment by lobpcg** [comment]
This PR is ready for review & merge, please.

**Comment by rgommers** [comment]
This is looking quite good to me now. The extensive set of docs added are great - maybe the one thing to change is make the links into proper references with a title, it's not clear where they are pointing to now. The few tweaks look like a nice generalization that make things more robust.

I think we should indeed get this in before the 1.10.x branch point. I don't have much time right now, but will circle back to this if no one else does by the end of next week.

**Comment by lobpcg** [comment]
> one thing to change is make the links into proper references with a title, it's not clear where they are pointing to now. 

@rgommers - Done.

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 1]
```diff
@@ -209,48 +211,238 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
...
 
-    >>> np.allclose(s3, s)
+    >>> np.allclose(s5, s)
     True
-    >>> np.allclose(np.abs(u3), np.abs(u.toarray()))
-    True
-    >>> np.allclose(np.abs(vT3), np.abs(vT.toarray()))
+
+    Since the singular values are not close to each other in this example,
+    every singular vector matchs as expected up to a difference in sign.
```
Comment (tylerjereddy) [inline scipy/sparse/linalg/_eigen/_svds.py:248]: `matchs` -? `matches` I think

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 2]
```diff
@@ -209,48 +211,238 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
...
+    are not so small.
+
+    >>> s_a(u2[:, :1], u[:, :1]) + s_a(u2[:, 1:], u[:, 1:]) > 1e-3
+    True
+
+    >>> (s_a(vT2[:1, :].T, vT[:1, :].T) +
+    ...  s_a(vT2[1:, :].T, vT[1:, :].T)) > 1e-3
+    True
+
+    As opposed to the angles between the 2-dimentional invariant subspaces
```
Comment (tylerjereddy) [inline scipy/sparse/linalg/_eigen/_svds.py:287]: dimentional -> dimensional I think

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 3]
```diff
@@ -209,48 +211,238 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
...
+    True
+
+    as well as for left singular vectors.
+
+    >>> s_a(vT2.T, vT.T).sum() < 1e-6
+    True
+
+    The next example follows that of 'sklearn.decomposition.TruncatedSVD'.
+
+    >>> rng = np.random.RandomState(0)
```
Comment (tylerjereddy) [inline scipy/sparse/linalg/_eigen/_svds.py:300]: isn't `RandomState` for legacy NumPy random generators?

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 4]
```diff
@@ -209,48 +211,238 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
...
+    >>> X_dense = rng.random(size=(100, 100))
+    >>> X_dense[:, 2 * np.arange(50)] = 0
+    >>> X = csr_matrix(X_dense)
+    >>> _, singular_values, _ = svds(X, k=5)
+    >>> print(singular_values)
+    [ 4.3293...  4.4491...  4.5420...  4.5987... 35.2410...]
+
+    The function can be called without the transpose of the input matrix
+    ever explicitely constructed.
```
Comment (tylerjereddy) [inline scipy/sparse/linalg/_eigen/_svds.py:309]: explicitely -> explicitly

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 5]
```diff
@@ -209,48 +211,238 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
...
+
+    >>> from scipy.linalg import svd
+    >>> from scipy.sparse import rand
+    >>> from scipy.sparse.linalg import aslinearoperator
+    >>> rng = np.random.RandomState(0)
```
Comment (tylerjereddy) [inline scipy/sparse/linalg/_eigen/_svds.py:314]: same here--legacy I think

File: scipy/sparse/linalg/_eigen/_svds.py [hunk 6]
```diff
@@ -209,48 +211,238 @@ def svds(A, k=6, ncv=None, tol=0, which='LM', v0=None,
...
+
+    >>> n = 5  # The dimension of the space.
+    >>> M_from_diff0 = diff0(np.eye(n))
+    >>> print(M_from_diff0.astype(int))
+    [[-1  1  0  0  0]
+     [ 0 -1  1  0  0]
+     [ 0  0 -1  1  0]
+     [ 0  0  0 -1  1]]
+
+    The matrix 'M_from_diff0' is bi-diagonasl and could be alternatively
```
Comment (tylerjereddy) [inline scipy/sparse/linalg/_eigen/_svds.py:341]: `diagonasl` -> `diagonal`

**Review by tylerjereddy — COMMENTED** [review]
Ralf seemed reasonably happy, and you addressed the oustanding comment from his review. I also read through the full diff and added a few comments.

However, most of my comments are sufficiently minor that I think I'll prefer to squash merge this in right now. CI is passing and the diff is fairly large, so probably easier to get this off my "release checklist" since it is in pretty good shape, and if you want to do a smaller follow-up my guess is that the legacy NumPy random infra might be a thing to fix up.

I'll note that I'm certainly not a domain expert, and the lack of such experts with review bandwidth is probably why these PRs sit without review sometimes. Still, the actual code changes seem pretty small and the testing tweaks seem to be mostly improvements. Some skipping changes maybe make me a little nervous, but sounds like this stuff causes problems even on `main` sometimes...

File: scipy/sparse/linalg/_eigen/_svds.py (hunk 3 above)
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/_svds.py:300]: This example just follows verbatim 'sklearn.decomposition.TruncatedSVD'

File: scipy/sparse/linalg/_eigen/_svds.py (hunk 5 above)
Comment (lobpcg) [reply scipy/sparse/linalg/_eigen/_svds.py:314]: This example just follows verbatim 'sklearn.decomposition.TruncatedSVD'

**Pull request merged**